ResearchKit/Common/ORKHeightPicker.m

```diff
@@ -0,0 +1,272 @@
+ the copyright holders even if such marks are included in this software.
+ 
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#import "ORKHeightPicker.h"
+#import "ORKResult_Private.h"
+#import "ORKAnswerFormat_Internal.h"
+#import "ORKAccessibilityFunctions.h"
+
+
+@interface ORKHeightPicker () <UIPickerViewDataSource, UIPickerViewDelegate>
+
+@end
+
+
+@implementation ORKHeightPicker {
+    UIPickerView *_pickerView;
+    ORKHeightAnswerFormat *_answerFormat;
+    id _answer;
+    __weak id<ORKPickerDelegate> _pickerDelegate;
+}
+
+@synthesize pickerDelegate = _pickerDelegate;
+
+- (instancetype)initWithAnswerFormat:(ORKHeightAnswerFormat *)answerFormat answer:(id)answer pickerDelegate:(id<ORKPickerDelegate>)delegate {
+    self = [super init];
+    if (self) {
+        NSAssert([answerFormat isKindOfClass:[ORKHeightAnswerFormat class]], @"answerFormat should be ORKHeightAnswerFormat");
+        
+        // Take into account locale changes so unit string can be updated
+        [[NSNotificationCenter defaultCenter] addObserver:self
+                                                 selector:@selector(currentLocaleDidChange:)
+                                                     name:NSCurrentLocaleDidChangeNotification object:nil];
+
+        self.answer = answer;
+        _answerFormat = answerFormat;
+        _pickerDelegate = delegate;
+    }
+    return self;
+}
+
+- (UIView *)pickerView {
+    if (_pickerView == nil) {
+        _pickerView = [[UIPickerView alloc] init];
+        _pickerView.dataSource = self;
+        _pickerView.delegate = self;
+        [self setAnswer:_answer];
+    }
+    return _pickerView;
+}
+
+- (void)setAnswer:(id)answer {
+    _answer = answer;
+    
+    if (ORKIsAnswerEmpty(answer)) {
+        answer = [self defaultAnswerValue];
+    }
+    
+    if (_answerFormat.useMetricSystem) {
+        NSUInteger index = [[self centimeterValues] indexOfObject:answer];
+        if (index == NSNotFound) {
+            [self setAnswer:[self defaultAnswerValue]];
+            return;
+        }
+        [_pickerView selectRow:index inComponent:0 animated:NO];
+    } else {
+        double feet, inches;
+        ORKInchesToFeetAndInches(((NSNumber *)answer).doubleValue, &feet, &inches);
```

Comment:
Same reason as below.

The answer value is used to look for the selected row in the picker (selectedRowInComponent:), so I'm afraid that converting back and forth to centimeters might introduce floating point rounding errors.

Because of this, I figured out that it's better to use the locale format in the picker (which when are in inches it should not cause rounding errors), and just convert to cm when then answer result is calculated.